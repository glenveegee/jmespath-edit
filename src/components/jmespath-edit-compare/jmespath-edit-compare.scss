:host {
  /* Component styles go here */
  padding: 1%;
  width: 98%;
  display: block;

  section {

    flex: 1;
    & h2 {
      font-family: sans-serif;

      > span {
        display: block;
        font-size: 0.8rem;
        color: #666;

        button {
          position: absolute;
          right: 2rem;
          margin-top: -1rem;
        }
      }
    }

    & div * {
      font-family: Consolas, Monaco, Menlo, 'Courier New', monospace;
      font-size: 12px;
    }
  }

  & .expression div {
    margin-bottom: 20px;
    width: inherit;

    & input, textarea {
      width: 100%;
      padding: 1%;
      box-sizing: border-box;
    }
  }

  & .results {
    display: flex;
    flex: 1 1 auto;
    flex-wrap: wrap;
    flex-direction: row;

    & section {
      > div {
        display: flex;
        flex: 1;
        flex-direction: column;
        min-width: 24rem;

        > .messages {
          display: flex;
          flex: 1;
          flex-direction: row;
          justify-content: flex-end;
          padding-right: 1rem;

          & .inputWarning {
            padding: 0rem 0 0.3rem 1rem;
            border-left: 4px solid goldenrod;
            color: goldenrod;
            max-height: 1rem;
            flex: 1;
          }

          & button {
            margin: 0.2rem 0.5rem;
          }
        }

        > .document-view {
          // max-width: min-content;
          // min-width: max-content;
          // padding: 0 0.5rem;
          width: calc(100% - 0.5rem);
          max-width: calc(100% - 0.5rem);
          height: 60vh;
          box-sizing: border-box;

          & pre,
          & textarea {
            -webkit-box-sizing: border-box;
            -moz-box-sizing: border-box;
            box-sizing: border-box;
            overflow: auto;
            width: inherit;
            height: inherit;
            padding: 0.5rem;
            white-space: pre-wrap;
          }
        }
      }

    }
  }

  & .output div.document-view {
    // padding: 1vw;
    background-color: #eee;
    // box-sizing: content-box;
    max-width: 100%;
    overflow: hidden;
    white-space: pre-wrap;
    word-break: break-all;
  }
}
