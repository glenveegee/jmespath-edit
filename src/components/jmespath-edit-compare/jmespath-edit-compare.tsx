import { Component, ComponentInterface, State, Host, h, Prop, Watch } from '@stencil/core';

import {BehaviorSubject, combineLatest, Subscription} from 'rxjs';
import {query as JMESPathQuery} from '../../utils/jmespath';
import {query as JMESPathPlusQuery} from '../../utils/jmespath-plus';
import {query as MetrichorJMESPathPlusQuery} from '../../utils/metrichor/jmespath-plus';
import {query as MetrichorJMESPathQuery} from '../../utils/metrichor/jmespath';
import { JSONValue } from '@metrichor/jmespath/dist/types';

@Component({
  styleUrl: 'jmespath-edit-compare.scss',
  tag: 'jmespath-edit-compare',
})
export class JmespathEdit implements ComponentInterface {

  library$ = new BehaviorSubject<string>('');
  expression$ = new BehaviorSubject<string>('');
  source$ = new BehaviorSubject<any>(null);
  query$ = combineLatest([this.expression$, this.source$, this.library$]);

  listener!: Subscription;
  @Prop() library = '@metrichor/jmespath-plus';

  @Prop() expression = '';
  @Prop() json: JSONValue = null;

  @State() output = '';
  @State() inputError = '';

  @Watch('library')
  updateLibraryHandler(newLibrary: string) {
    if (newLibrary) {
      this.library$.next(newLibrary)
    }
  }

  @Watch('expression')
  updateExpressionHandler(newExpression: string) {
    if (newExpression) {
      this.expression$.next(newExpression)
    }
  }

  @Watch('json')
  updateSourceHandler(newSource: JSONValue) {
    if (newSource) {
      this.source$.next(newSource)
    }
  }

  runLibrarySpecificQuery = async (expression: string, source: JSONValue, library: string): Promise<JSONValue> => {
    switch (library) {
      case 'jmespath':
        return JMESPathQuery(expression, source);
      case '@metrichor/jmespath-plus':
        return MetrichorJMESPathPlusQuery(expression, source)
      case '@metrichor/jmespath':
        return MetrichorJMESPathQuery(expression, source)
      case 'jmespath-plus':
        return JMESPathPlusQuery(expression, source)
      default:
        return JMESPathQuery(expression, source);
    }
  }

  runQuery = async ([expression, source, library]: [string, JSONValue, string]) => {
    if (!expression || !source) return
    try {
      const result = await this.runLibrarySpecificQuery(expression, source, library);
      this.output = JSON.stringify(result, null, 2)
    } catch (error) {
      this.output = error.message
    }
  }

  coerceJSON = (json: JSONValue): JSONValue => {
    if (!json) return json;
    if (typeof json === 'string') {
      try {
        return JSON.parse(json)
      } catch (error) {
        return json
      }
    }
    return json;
  }

  componentWillLoad() {
    this.listener = this.query$.subscribe(this.runQuery)
    this.expression$.next(this.expression);
    this.source$.next(this.coerceJSON(this.json));
    this.library$.next(this.library);
  }

  disconnectedCallback() {
    this.listener.unsubscribe()
  }

  setSource = (e: any) => {
    const source = e.target.value;
    if (!source) return;
    try {
      this.source$.next(JSON.parse(source))
      this.inputError = '';
    } catch (error) {
      console.error(error)
      this.inputError = error.message
    }
  }

  setExpression = (e: any) => {
    const expression = e.target.value;
    if (expression) {
      this.expression$.next(expression)
    }
  }

  render() {

    const currentExpression = this.expression$.getValue();
    const currentSource = this.source$.getValue();

    return (
      <Host>
        <section class="expression">
          <h2>EXPRESSION</h2>
          <div>
            <input type="text" value={currentExpression}
            onInput={this.setExpression}
            />
          </div>
        </section>
        <div class="results">
          <section class="input">
            <h2>SOURCE DATA</h2>
            <div>
              {
                this.inputError && <div class='inputWarning'>{this.inputError}</div> || null
              }
              <textarea value={currentSource && JSON.stringify(currentSource, null, 2) || ''} onInput={this.setSource} />
            </div>
          </section>
          <section class="output">
            <h2>OUTPUT</h2>
            <div>
              <pre>{this.output}</pre>
            </div>
          </section>
        </div>
      </Host>
    );
  }
}
