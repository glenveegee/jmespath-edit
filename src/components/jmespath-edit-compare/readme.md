# jmespath-edit-compare



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute    | Description | Type                                                                                     | Default                      |
| ------------ | ------------ | ----------- | ---------------------------------------------------------------------------------------- | ---------------------------- |
| `expression` | `expression` |             | `string`                                                                                 | `''`                         |
| `json`       | `json`       |             | `JSONValue[] \| boolean \| null \| number \| string \| { [member: string]: JSONValue; }` | `null`                       |
| `library`    | `library`    |             | `string`                                                                                 | `'@metrichor/jmespath-plus'` |


## Dependencies

### Used by

 - [jmespath-edit-demo](../jmespath-edit-demo)

### Graph
```mermaid
graph TD;
  jmespath-edit-demo --> jmespath-edit-compare
  style jmespath-edit-compare fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
