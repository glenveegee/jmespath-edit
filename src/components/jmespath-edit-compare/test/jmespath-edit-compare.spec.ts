import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { JmespathEdit } from '../jmespath-edit-compare';

describe('jmespath-edit-compare', () => {
  // let rootInst: JmespathEdit;
  let rootEl: HTMLJmespathEditElement;
  let page: SpecPage;

  beforeEach(async () => {
    page = await newSpecPage({
      components: [JmespathEdit],
      html: '<jmespath-edit-compare></jmespath-edit-compare>',
    });
    // rootInst = page.rootInstance;
    rootEl = page.root as HTMLJmespathEditElement;
  });

  describe('render', () => {
    it('renders default datasource and query', () => {
      expect((rootEl.shadowRoot.querySelector('.expression div textarea') as HTMLTextAreaElement)).toEqualHtml('<textarea value=\"\"></textarea>')
      expect((rootEl.shadowRoot.querySelector('.results .input div textarea') as HTMLTextAreaElement)).toEqualLightHtml('<textarea value=\"\"></textarea>')
      expect((rootEl.shadowRoot.querySelector('.results .output div pre') as HTMLPreElement)).toEqualText('')
    });

    it('renders default datasource and query', async () => {
      rootEl.expression = "locations[?state == 'WA'].name | sort(@) | {WashingtonCities: join(', ', @)}";
      rootEl.json = {"locations": [{"name": "Seattle", "state": "WA"},{"name": "New York", "state": "NY"},{"name": "Bellevue", "state": "WA"},{"name": "Olympia", "state": "WA"}]};
      await page.waitForChanges();

      expect(((rootEl.shadowRoot.querySelector('.expression > div > textarea') as HTMLTextAreaElement))).toEqualHtml(`<textarea value=\"locations[?state == 'WA'].name | sort(@) | {WashingtonCities: join(', ', @)}\"></textarea>`)
      expect((rootEl.shadowRoot.querySelector('.results .input div textarea') as HTMLTextAreaElement)).toEqualLightHtml(`<textarea value=\"{
  &quot;locations&quot;: [
    {
      &quot;name&quot;: &quot;Seattle&quot;,
      &quot;state&quot;: &quot;WA&quot;
    },
    {
      &quot;name&quot;: &quot;New York&quot;,
      &quot;state&quot;: &quot;NY&quot;
    },
    {
      &quot;name&quot;: &quot;Bellevue&quot;,
      &quot;state&quot;: &quot;WA&quot;
    },
    {
      &quot;name&quot;: &quot;Olympia&quot;,
      &quot;state&quot;: &quot;WA&quot;
    }
  ]
}\"></textarea>`)
      expect((rootEl.shadowRoot.querySelector('.results .output div pre') as HTMLPreElement)).toEqualText(`
{
  \"WashingtonCities\": \"Bellevue, Olympia, Seattle\"
}`)
    });
  });
});
