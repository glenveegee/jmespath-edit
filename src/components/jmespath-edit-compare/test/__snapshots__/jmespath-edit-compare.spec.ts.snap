// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jmespath-edit-compare render renders default datasource and query 1`] = `<pre></pre>`;
